### PR TITLE
supply local time to InternalPushTxn when initiating the request

### PR DESCRIPTION
**Pull request by tbg**
previously args.Timestamp was used, but there are situations
in which that never changes, meaning that infinite retry loops
on an abandoned Txn descriptor would ensue in storage/store,
for instance in #877 and the bank example.

This needs a test. Should be fairly straightforward, so if anybody wants to help out the debugging bonanza, that would be appreciated (but probably want to wait for feedback to see if this would actually land). 

After this change, the following hums along relatively nicely on a three node cluster:

`/bank --num-accounts=2 --num-parallel-transfers=2 --db-name=http://root@localhost:8001`

Every couple of minutes one of the writers' transactions will still deadlock, but times out after
10 seconds (after which process resumes). When that happens, the busy retry loop will run against it
approximately 14000 times until the 10s are over (yeah, it's really busy - 0s backoff, I think). I've checked all the pushes in such an example, and none of them go through since the priority is pretty high:

```
# count, pusher prio, pushee prio
6297 21781967 < 2132569132
   1 271073717 < 2132569132
3816 1372547470 < 2132569132
3815 2132569131 < 2132569132
```

So we only have to figure out why the occasional transaction in the bank example doesn't finish - it clearly seems to be pretty high priority wise. I've checked the logs, it doesn't seem to ever have unsuccessfully pushed someone during my run:

```
6297 21781967 < 2132569132
   1 271073717 < 2132569132
3816 1372547470 < 2132569132
3815 2132569131 < 2132569132
```

### CODE REVIEW

File: storage/range_command.go [hunk 1]
```diff
@@ -567,10 +567,11 @@ func (r *Range) InternalPushTxn(batch engine.Engine, ms *proto.MVCCStats, args *
 	if reply.PusheeTxn.LastHeartbeat == nil {
 		reply.PusheeTxn.LastHeartbeat = &reply.PusheeTxn.Timestamp
 	}
-	// Compute heartbeat expiration (use args.Timestamp instead of node
-	// clock wall time so that all replicas see the same result).
-	expiry := args.Timestamp
+	// Compute heartbeat expiration (all replicas must see the same result).
+	expiry := args.Now             // caller can set this to his wall time.
+	expiry.Forward(args.Timestamp) // if Now is not set, fallback
```
Comment (bdarnell) [inline storage/range_command.go]: Is there any reason to send a push txn request without setting `Now`? It might be simpler to require that `Now` be set in all cases. (On the other hand, this does ensure that we use `max(Now, Timestamp)` if there is any risk that the pusher's clock could be behind)

File: proto/internal.proto [hunk 2]
```diff
@@ -121,17 +121,23 @@ enum PushTxnType {
 message InternalPushTxnRequest {
   optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
   optional Transaction pushee_txn = 2 [(gogoproto.nullable) = false];
+  // Now optionally holds the timestamp used to compare the last heartbeat
```
Comment (bdarnell) [inline proto/internal.proto]: I found this comment hard to follow since it talks in terms of the low-level mechanism. How about:

Now is the time when this request was issued (captured in the request by the sender so all replicas have a consistent view of time). The current time is used instead of the transaction's timestamp in order to handle retries and detect abandoned transactions.

**Comment by bdarnell** [comment]
LGTM. Do we also need to set Now when pushing from gc_queue.go? (which looks like the only other non-test use of InternalPushTxnRequest)

**Comment by tschottdorf** [comment]
I've made `Now` mandatory and updated the code. PTAL.

File: storage/range_test.go [hunk 3]
```diff
@@ -1766,9 +1767,9 @@ func TestInternalPushTxnHeartbeatTimeout(t *testing.T) {
 		pushType    proto.PushTxnType
 		expSuccess  bool
 	}{
-		{nil, 0, proto.PUSH_TIMESTAMP, false},
-		{nil, 0, proto.ABORT_TXN, false},
-		{nil, 0, proto.CLEANUP_TXN, false},
+		{nil, 1, proto.PUSH_TIMESTAMP, false}, // using 0 is awkward
```
Comment (bdarnell) [inline storage/range_test.go]: This comment won't make sense in the future: using 0 for what?

File: storage/range_test.go [hunk 4]
```diff
@@ -1807,7 +1808,8 @@ func TestInternalPushTxnHeartbeatTimeout(t *testing.T) {
 		// Now, attempt to push the transaction with clock set to "currentTime".
 		tc.manualClock.Set(test.currentTime)
 		args, reply := pushTxnArgs(pusher, pushee, test.pushType, 1, tc.store.StoreID())
-		args.Timestamp = tc.clock.Now()
+		args.Timestamp = proto.Timestamp{WallTime: test.currentTime} // avoid logical ticks
```
Comment (bdarnell) [inline storage/range_test.go]: Why is it important to avoid logical ticks?

**Comment by bdarnell** [comment]
LGTM

File: storage/range_test.go (hunk 4 above)
Comment (tschottdorf) [inline storage/range_test.go]: background processes will add up logical ticks and so the comparisons are random, failing the borderline cases.

Comment (tschottdorf) [inline storage/range_test.go]: I'll add that to the comment.

File: storage/range_test.go (hunk 3 above)
Comment (tschottdorf) [inline storage/range_test.go]: added to the comment.
